### PR TITLE
Cosmos-3158 DateSelector component

### PR DESCRIPTION
**Pull request by ivyyangthl**
1. Functionalities working
2. Prepopulate data, generate output
3. Rebass theme
4. Validation included
5. index.md included

### CODE REVIEW

File: RebassThemes/index.jsx [hunk 1]
```diff
@@ -18,13 +18,21 @@ export default class CosmosComponents extends React.Component {
 			rebass: {
 				...rebassConfig,
 				Button: {
```
Comment (SimeonC) [inline RebassThemes/index.jsx]: This file should not be changed as part of the PR. Changing this has 2 fold effect; 1. It'll probably be overridden in the implementation. 2. You're putting specific styling into a generic base style - this isn't something we want to control or rely on at this level.

File: DateSelector/index.jsx [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+		rebass: React.PropTypes.object
+	};
+
+	constructor(props) {
+		super(props);
+
+		const startTime = convertTimeFromMomentObj(this.props.startDateTime);
+		const endTime = convertTimeFromMomentObj(this.props.endDateTime);
+
+		this.state = {
```
Comment (SimeonC) [inline DateSelector/index.jsx]: The timeSelector is correctly written, why did you revert to using state for data in this one?
This component has no need or requirement to interface with the state for its data storage.

File: DateSelector/index.jsx [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+
+			hourBefore: (endTime === null? 11 : endTime.hour) + '',  // inital value for Before
+			minutesBefore: (endTime === null? 59 : endTime.minutes) + '',
+			isAMBefore: endTime === null? false : endTime.isAM,
+
+			repeatOption: this.props.repeatOption || 'Does Not Repeat',
+			repeatUntil: this.props.repeatUntil || null,
+			focusedUntil: false,
+
+			Mon: this.props.Mon === undefined? false : this.props.Mon,   // default value for day selection
```
Comment (SimeonC) [inline DateSelector/index.jsx]: 2 Problems here: Capitalised properties are wrong, should always be camel-case. (`isAMBefore` is a grey area) Also, in this case, you should really be using an array of strings, constants and Array.indexOf not individual booleans.

File: DateSelector/index.jsx [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+
+				dateMode: this.state.dateMode,
+			};
+		}
+
+		const errorMessage = validateOutput(data);
+		this.setState({message: errorMessage});
+
+		if(errorMessage.length === 0){
+			this.props.addCalendar(data);
```
Comment (SimeonC) [inline DateSelector/index.jsx]: The reason doing this is a bad idea is due to React's way of checking for changes - if you change the object it will rerender all other components that use a key on the object.

By passing and changing individual properties the component becomes more reusable and validation/restrictions becomes more customisable.

File: DateSelector/index.jsx [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+								isAMAfter: isAM
+							})}
+						/>
+					</div>
+				}
+
+				{
+					(this.state.dateMode !== 'After') &&
+					<div>
+						<Label>Before</Label>
```
Comment (SimeonC) [inline DateSelector/index.jsx]: String labels should be passed through on props with defaults set to this value - otherwise, they cannot be translated later. (or generally modified by implementation)

File: DateSelector/index.jsx (hunk 2 above)
Comment (ivyyangthl) [inline DateSelector/index.jsx]: This comes from my traditional 'object-oriented' way of thinking: wanting to encapsulate the logic that the outside world does not care inside the UI component, define easy-to-use in/out interface. This is so different from React way of doing things

File: DateSelector/index.jsx (hunk 5 above)
Comment (ivyyangthl) [inline DateSelector/index.jsx]: Yes, should take care of intl

File: RebassThemes/index.jsx (hunk 1 above)
Comment (ivyyangthl) [inline RebassThemes/index.jsx]: Shall we remove all the specific styling from this file?

Comment (SimeonC) [inline RebassThemes/index.jsx]: No, the styling there is there for a reason. Mostly. I think the Button ones actually get overriden by the component anyway so those could probably go. But the Label style will have unexpected knockon effects.
Also the ButtonOutline component sets the border colour by it's "theme" property.

File: DateSelector/index.jsx (hunk 2 above)
Comment (SimeonC) [inline DateSelector/index.jsx]: Javascript is Functional programming at its heart. It just resembles OOP sometimes.

File: DateSelector/index.jsx (hunk 4 above)
Comment (SimeonC) [inline DateSelector/index.jsx]: You may need some state, the idea is to use as absolute little of the state as possible. My introduction to state was - State Bad, Props Good.

Comment (ivyyangthl) [inline DateSelector/index.jsx]: This makes sense - same as 'Step 4: Identify Where Your State Should Live' on https://facebook.github.io/react/docs/thinking-in-react.html

File: RebassThemes/index.jsx (hunk 1 above)
Comment (SimeonC) [inline RebassThemes/index.jsx]: Oh I see, the diff showed up weird - sorry. Looking at it, NavigationToolbar is the only one that should be there as that's for one of our components. Button and ButtonOutline shouldn't be there but that should be a different PR as we'd have to go and check all our cosmos projects and check it didn't break anything.

Comment (ivyyangthl) [inline RebassThemes/index.jsx]: Sounds good now
